packages/flutter/lib/src/material/time_picker.dart

```diff
@@ -1738,9 +1738,7 @@ class _TimePickerInputState extends State<_TimePickerInput> with RestorationMixi
 
   String? _validateHour(String? value) {
     final int? newHour = _parseHour(value);
-    setState(() {
-      hourHasError.value = newHour == null;
-    });
+    hourHasError.value = newHour == null;
```

Comment:
This looks like a clear improvement: [`FormState.validate`](https://main-api.flutter.dev/flutter/widgets/FormState/validate.html) guarantees a rebuild, so I don't believe this `setState` is necessary.

<br>

However: I'm a bit worried about the implications here.

There might be a bunch Flutter projects outside of this codebase that build a `Form` with a similar structure; sadly, migrating the code here isn't going to help them out.

(I personally feel that many different aspects of the `Form` widget could be improved, but they'd require significant time investment and would definitely qualify as "breaking changes". I've also found that it's not very hard to just ignore `Form` entirely by using a state management package like provider or riverpod.)

<br>

Assuming we go through with this change: much like how theme data has [`useMaterial3`](https://main-api.flutter.dev/flutter/material/ThemeData/useMaterial3.html), it'd be nice if `Form` had something that allowed for a migration period.

Reply:
I've been eyeing Form for a rewrite for a long time. I wish I could just write a whole new class with a new name, but the name "Form" is so good I think we'd have to do something like userMaterial3.